### PR TITLE
Conditionally use flashes in failure_app

### LINKED ISSUE #3574: Flash support in warden failures
I'm using json as the only endpoint and rails-api gem.
I'm getting exception `undefined method 'flash'` in login error responses, looking at the code like https://github.com/plataformatec/devise/blob/master/lib/devise/failure_app.rb#L52 it seems it's not conditionally use flash as in controllers https://github.com/plataformatec/devise/blob/master/app/controllers/devise/sessions_controller.rb#L18 only when the format is navigational, maybe it should be checked too in the failure_app?

### PR DESCRIPTION
**Pull request by alex88**
Fixes #3574

### CODE REVIEW

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by serene** [comment]
@alex88 Would we then be able to get error messages from somewhere else, if flash messages are no longer being populated in flash?

**Comment by alex88** [comment]
@serene you already get the error In the response afaik

**Comment by serene** [comment]
@alex88 Would this be the case even in the event of a remote/JS sign in call?

**Comment by alex88** [comment]
@serene I think so, I don't remember the case, but I think I was exactly calling the API to sign in via ajax

File: lib/devise/failure_app.rb [hunk 1]
```diff
@@ -49,17 +49,19 @@ def http_auth
 
     def recall
       env["PATH_INFO"]  = attempted_path
-      flash.now[:alert] = i18n_message(:invalid)
+      flash.now[:alert] = i18n_message(:invalid) if is_flashing_format?
```
Comment (serene) [inline lib/devise/failure_app.rb]: I think the errors are getting lost for ajax calls. They never get set anywhere else for a non-navigational format.

Comment (alex88) [inline lib/devise/failure_app.rb]: Have you tried? What's the response of the ajax call?

**Pull request merged**